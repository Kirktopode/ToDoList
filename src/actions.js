/*
 * action types
 */

export const ADD_TODO = 'ADD_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';
export const SET_VISIBILITY_FILTER = 'SET_VISIBILITY_FILTER';
export const TOGGLE_DISPLAY_TIME = 'TOGGLE_DISPLAY_TIME'

/*
 * other constants
 */

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE'
}

/*
 * action creators
 */

export function addTodo(text, time){
  return{
    type: ADD_TODO,
    text: text,
    time: time
  };
}

export function toggleTodo(index){
  return{
    type: TOGGLE_TODO,
    index: index
  };
}

export function setVisibilityFilter(filter){
  return{
    type: SET_VISIBILITY_FILTER,
    filter: filter
  }
}

export function toggleDisplayTime(){
  return{
    type: TOGGLE_DISPLAY_TIME
  }
}
