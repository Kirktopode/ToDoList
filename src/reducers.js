import {
  ADD_TODO,
  DELETE_TODO,
  COPY_TODO,
  APPEND_TIMESTAMP,
  VisibilityFilters,
  TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  TOGGLE_DISPLAY_TIME
} from './actions'
import {
  dateString,
  appendDateString
} from './timeModule'
import { combineReducers } from 'redux'

const { SHOW_ALL } = VisibilityFilters;

const initialState = {
  visibilityFilter: VisibilityFilters.SHOW_ALL,
  todos: [],
  displayTime: true
}

const todoApp = combineReducers({
  visibilityFilter,
  todos,
  displayTime
})

export default todoApp;

function visibilityFilter(state = SHOW_ALL, action) {
  switch(action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

function displayTime(state = true, action){
  switch(action.type){
    case TOGGLE_DISPLAY_TIME:
      //console.log("toggle")
      //console.log(state)
      return !state
    default:
      return state
  }
}

function todos(state = [], action){
  switch(action.type){
    case ADD_TODO:
      return [
        ...state,
        {
          text: action.text,
          time: dateString(),
          completed: false,
          displayTime: true
        }
      ].map((todo, index) => {
        return Object.assign({}, todo, {
          key: index
        })
      })
    case DELETE_TODO:
      return state.filter((todo, index) => {
        return index != action.index
      }).map((todo, index) => {
        return Object.assign({}, todo, {
          key: index
        })
      })
    case COPY_TODO:
      let copies = []
      for(let i = 0; i < action.copies; i++){
        copies.push({
          text: state[action.index].text,
          time: dateString(),
          completed: false,
          displayTime: true
        })
      }
      return [
        ...state,
        ...copies
      ].map((todo, index) => {
        return Object.assign({}, todo, {
          key: index
        })
      })
    case APPEND_TIMESTAMP:
      return state.map((todo, index) => {
        if(index == action.index){
          return Object.assign({}, todo, {
            time: todo.time + appendDateString()
          })
        }
        return todo
      })
    case TOGGLE_TODO:
      return state.map((todo, index) => {
        if(index === action.index){
          return Object.assign({}, todo, {
            completed: !todo.completed
          })
        }
        return todo;
      })
    default:
      return state;
  }
}
