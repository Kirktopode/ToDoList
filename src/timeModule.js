export function dateString(){
  var date = new Date()
  return " - Created " + date.toLocaleTimeString() + " " + date.toLocaleDateString()
}
export function appendDateString(){
  var date = new Date()
  return " - Time Appended " + date.toLocaleTimeString() + " " + date.toLocaleDateString()
}
