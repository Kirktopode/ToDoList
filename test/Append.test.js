import React from 'react'
import { render } from 'react-dom'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { addTodo } from '../src/actions'
import todoApp from '../src/reducers'
import App from '../src/components/App'
import renderer from 'react-test-renderer';

test('Timestamp is appended properly when the button is clicked.', () =>{

  //Mocking Date for the sake of testing
  global.Date = jest.fn(() => {return {
    toLocaleDateString: () => {
      return "test date"
    },
    toLocaleTimeString: () => {
      return "test time"
    }
  }})

  let store = createStore(todoApp)
  store.dispatch(addTodo("Test"))

  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>
  );

  let tree = component.toJSON()
  expect(tree.children[1].children[0].children[2].children[0]).toMatch(/Created test time test date$/)
  expect(tree).toMatchSnapshot()

  //Append timestamp
  tree.children[1].children[0].children[1].props.onClick()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(tree.children[1].children[0].children[2].children[0]).toMatch(/Appended test time test date$/)
})
