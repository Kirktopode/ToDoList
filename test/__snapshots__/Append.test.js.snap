// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timestamp is appended properly when the button is clicked. 1`] = `
<div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <input />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
  </div>
  <ul>
    <li
      style={
        Object {
          "textDecoration": "none",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        x
      </button>
      <button
        onClick={[Function]}
      >
        Append Time
      </button>
      <span
        onClick={[Function]}
      >
        Test  - Created test time test date
      </span>
      <input
        id="copy-count-0"
        type="number"
      />
      <button
        onClick={[Function]}
      >
        Copy
      </button>
    </li>
  </ul>
  <p>
    Show:
     
    <span>
      All
    </span>
    , 
    <a
      href=""
      onClick={[Function]}
    >
      Active
    </a>
    , 
    <a
      href=""
      onClick={[Function]}
    >
      Completed
    </a>
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    Time
  </p>
</div>
`;

exports[`Timestamp is appended properly when the button is clicked. 2`] = `
<div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <input />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
  </div>
  <ul>
    <li
      style={
        Object {
          "textDecoration": "none",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        x
      </button>
      <button
        onClick={[Function]}
      >
        Append Time
      </button>
      <span
        onClick={[Function]}
      >
        Test  - Created test time test date - Time Appended test time test date
      </span>
      <input
        id="copy-count-0"
        type="number"
      />
      <button
        onClick={[Function]}
      >
        Copy
      </button>
    </li>
  </ul>
  <p>
    Show:
     
    <span>
      All
    </span>
    , 
    <a
      href=""
      onClick={[Function]}
    >
      Active
    </a>
    , 
    <a
      href=""
      onClick={[Function]}
    >
      Completed
    </a>
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    Time
  </p>
</div>
`;
